lib/srv/uacc/uacc_linux.go

```diff
@@ -234,6 +244,61 @@ func UserWithPtyInDatabase(utmpPath string, username string) error {
 	}
 }
 
+// LogFailedLogin writes a new entry to the btmp failed login log.
+// This should be called when an interactive session fails due to a missing
+// local user.
+//
+// `username`: Name of the user the interactive session is running under.
+// `hostname`: Name of the system the user is logged into.
+// `remote`: IPv6 address of the remote host.
+func LogFailedLogin(btmpPath, username, hostname string, remote [4]int32) error {
+	// String parameter validation.
+	if len(username) > userMaxLen {
+		return trace.BadParameter("username length exceeds OS limits")
+	}
+	if len(hostname) > hostMaxLen {
+		return trace.BadParameter("hostname length exceeds OS limits")
+	}
+
+	btmpPath = getDefaultBtmpPath(btmpPath)
+	// Convert Go strings into C strings that we can pass over ffi.
+	cBtmpPath := C.CString(btmpPath)
+	defer C.free(unsafe.Pointer(cBtmpPath))
+	cUsername := C.CString(username)
+	defer C.free(unsafe.Pointer(cUsername))
+	cHostname := C.CString(hostname)
+	defer C.free(unsafe.Pointer(cHostname))
+
+	// Convert IPv6 array into C integer format.
+	cIP := [4]C.int{0, 0, 0, 0}
+	for i := 0; i < 4; i++ {
+		cIP[i] = (C.int)(remote[i])
+	}
+
+	timestamp := time.Now()
+	secondsElapsed := (C.int32_t)(timestamp.Unix())
+	microsFraction := (C.int32_t)((timestamp.UnixNano() % int64(time.Second)) / int64(time.Microsecond))
+
+	accountDb.Lock()
+	defer accountDb.Unlock()
+
+	status, errno := C.uacc_add_btmp_entry(cBtmpPath, cUsername, cHostname, &cIP[0], secondsElapsed, microsFraction)
+	switch status {
+	case C.UACC_UTMP_MISSING_PERMISSIONS:
+		return trace.AccessDenied("missing permissions to write to btmp")
+	case C.UACC_UTMP_WRITE_ERROR:
+		return trace.AccessDenied("failed to add entry to btmp")
+	case C.UACC_UTMP_FAILED_OPEN:
+		return trace.AccessDenied("failed to open user account database, code: %d", errno)
+	case C.UACC_UTMP_FAILED_TO_SELECT_FILE:
+		return trace.BadParameter("failed to select file")
+	case C.UACC_UTMP_PATH_DOES_NOT_EXIST:
+		return trace.NotFound("user accounting files are missing from the system, running in a container?")
+	default:
+		return decodeUnknownError(int(status))
+	}
+}
```

Comment:
Will any of this materialize due to the final call in the `C.uacc_add_btmp_entry` function?

```c
    updwtmp(file, &entry);
    return 0;
```

In `uacc_add_btmp_entry`, we expect the caller to check `errno`, but in Go `LogFailedLogin` expects non-zero status in case of errors. Both sides of the call effectively expect the other to do the job.

Reply:
`check_abs_path_err` can generate these statuses, and it's called in `uacc_add_btmp_entry`.

Reply:
True, but later on we call `updwtmp` which may report error using `errno`. We should check it after the call, either in C or in Go. For consistency I think we should do it in C, similar to the code in `check_abs_path_err`.